### PR TITLE
New Edit File Tool

### PR DESCRIPTION
**Pull request by bassamsdata**
## Description

Add a File Edit tool with deterministic and intelligent matching strategies.

### Overview

Introduces the renewed `insert_edit_into_file`, a tool that allows LLMs to make reliable, deterministic file edits using structured function calls. It supports substring and block edits, full overwrites, and boundary insertions while maintaining atomic behavior, all edits succeed or none are applied.

### Key Features

* **Atomic operations** – applies all edits or none to prevent partial file corruption
* **Multiple matching strategies** – exact, whitespace-normalized, punctuation-tolerant, block-anchored, and substring-based
* **Substring mode** – efficient for short token replacements (max 1000 occurrences)
* **Boundary markers** – insert at file start (`^` or `<<START>>`) or end (`$` or `<<END>>`)
* **Complete overwrite mode** – replace entire file content efficiently
* **Error diagnostics** – context-aware messages with similarity suggestions and line-level feedback
* **Conflict detection** – prevents overlapping or incompatible edits
* **File limits** – up to 2MB files and 50KB search text per edit

### Complete JSON Schema

```json
{
  "filepath": "src/utils.js",
  "mode": "append",
  "dryRun": false,
  "explanation": "Refactor function names and improve logging",
  "edits": [
    {
      "oldText": "function calc(a, b) {\n  return a + b;\n}",
      "newText": "function calculate(first, second) {\n  return first + second;\n}"
    },
    {
      "oldText": "console.log('debug')",
      "newText": "logger.info('debug')",
      "replaceAll": true
    }
  ]
}
```

### Parameter Reference

Required:

* `filepath` – path to the target file
* `edits` – array of edit operations

Optional:

* `mode` – `"append"` (default) or `"overwrite"` (replace file)
* `dryRun` – preview changes before writing (`false` by default)
* `explanation` – brief text describing the edit

Per-edit options:

* `oldText` – text to find
* `newText` – replacement text
* `replaceAll` – replace all occurrences (`false` by default)

### Special Cases

Overwrite mode:

```json
{
  "filepath": "config.js",
  "mode": "overwrite",
  "edits": [
    {"oldText": "", "newText": "const config = { api: 'v2', debug: false };"}
  ]
}
```

File boundaries:

```json
{
  "filepath": "script.py",
  "edits": [
    {"oldText": "^", "newText": "#!/usr/bin/env python3\n"},
    {"oldText": "$", "newText": "\nprint('Done')"}
  ]
}
```

### Architecture Overview

1. **Entry Point** – separates substring and block edits, ensuring atomic application
2. **Sequential Processing** – substring edits handled in parallel, block edits sequentially
3. **Matching Chain** – tries multiple fallback strategies with confidence scoring:

   * exact_match (1.0)
   * substring_exact_match (1.0)
   * whitespace_normalized (0.95)
   * punctuation_normalized (0.93)
   * position_markers (1.0)
   * trimmed_lines (0.8)
   * block_anchor (0.6)
4. **Conflict Detection** – identifies overlapping or conflicting regions
5. **Error Reporting** – suggests near-matches, shows line numbers, and hints for fixable cases

### Testing Summary

Validated with over 20 LLMs, handling up to 100 edits per file and 1000-line test cases. Tested providers and models include:

* **Proprietary Models:** OpenAI Models including the old 4.1, Gemini, Grok XAI
* **Open Source Models:** Qwen 3 Coder, Kimi K2, Gym 4.6, Gym 4.5 Air, Owen 14B, Owen 30b and OpenAI OSS series

The tool correctly handled structured JSON, quoted JSON, and Python-like boolean syntax across all tested models.

### Status

✅ Stable and ready for general use.

## Related Issue(s)

[https://github.com/olimorris/codecompanion.nvim/discussions/1607#discussioncomment-14397658](https://github.com/olimorris/codecompanion.nvim/discussions/1607#discussioncomment-14397658)
#2128
#2029
#1619

## Checklist

- [x] I've read the [contributing](https://github.com/olimorris/codecompanion.nvim/blob/main/CONTRIBUTING.md) guidelines and have adhered to them in this PR
- [x] I've added [test](https://github.com/olimorris/codecompanion.nvim/blob/main/CONTRIBUTING.md#testing) coverage for this fix/feature
- [x] I've run `make all` to ensure docs are generated, tests pass and my formatting is applied
- [ ] _(optional)_ I've updated `CodeCompanion.has` in the [init.lua](https://github.com/olimorris/codecompanion.nvim/blob/main/lua/codecompanion/init.lua#L239) file for my new feature
- [ ] _(optional)_ I've updated the README and/or relevant docs pages

### CODE REVIEW

File: lua/codecompanion/config.lua [hunk 1]
```diff
@@ -64,12 +64,13 @@ local defaults = {
...
             prompt = "I'm giving you access to the ${tools} to help you perform coding tasks",
             tools = {
+              "edit_tool_exp",
               "cmd_runner",
               "create_file",
               "file_search",
               "get_changed_files",
               "grep_search",
-              "insert_edit_into_file",
+              -- "insert_edit_into_file",
```
Comment (olimorris) [inline lua/codecompanion/config.lua]: This should replace the `insert_edit_into_file` tool. So when this is ready, I propose we rename it to that

File: lua/codecompanion/strategies/chat/tools/catalog/edit_tool_exp.lua [hunk 2]
```diff
@@ -0,0 +1,1022 @@
...
+      else
+        return edit_file(args, self.chat.bufnr, output_handler, self.tool.opts)
+      end
+    end,
+  },
+  schema = {
+    type = "function",
+    ["function"] = {
+      name = "edit_tool_exp",
+      description = "Robustly edit files with multiple automatic fallback strategies. Handles whitespace differences, indentation variations, and ambiguous matches intelligently.",
```
Comment (olimorris) [inline lua/codecompanion/strategies/chat/tools/catalog/edit_tool_exp.lua]: Could we move the system prompt into the tool's description? I'm hoping to piece together an "agent" like mode very soon and really want to avoid tool specific usage in the system prompt.

File: tests/strategies/chat/tools/catalog/insert_edit_into_file/test_edit_file.lua [hunk 3]
```diff
@@ -0,0 +1,1028 @@
...
+    post_once = child.stop,
+  },
+})
+
+T["Basic Functionality"] = new_set()
+
+T["Basic Functionality"]["can edit a simple file"] = function()
+  child.lua([[
+    -- create initial file
+    local initial = "function getName() {\n  return 'John';\n}"
```
Comment (olimorris) [inline tests/strategies/chat/tools/catalog/insert_edit_into_file/test_edit_file.lua:42]: I propose that we put this "initial" or "old_text" in a stub file. It will make it easier in the future to test specific code snippets for specific filetypes

Could name them:

- `simple_edit.php.oldText`
- `simple_edit.php.newText`

and create a helper function like `create_stub("simple_edit.php")` which returns the contents of both.

I'm happy to own this btw.

**Review by olimorris — COMMENTED** [review]
Completely onboard with this approach. The strategy chain is a brilliant idea alongside the confidence levels. Also, your testing is as comprehensive as it could get. This will be an amazing feature.

Sorry if I made you come back to the laptop by carrying on that conversation. No pressure at all.

File: tests/strategies/chat/tools/catalog/insert_edit_into_file/test_edit_file.lua (hunk 3 above)
Comment (bassamsdata) [reply tests/strategies/chat/tools/catalog/insert_edit_into_file/test_edit_file.lua:42]: This is much better for future testing.

> I'm happy to own this btw.

Honestly, that would be fantastic if you have the time, thank you. If you don’t, no worries; I’ll take care of it.

Comment (olimorris) [reply tests/strategies/chat/tools/catalog/insert_edit_into_file/test_edit_file.lua:42]: Of course. It feels like a sensible paradigm to adopt for testing in the plugin anyway

File: lua/codecompanion/config.lua (hunk 1 above)
Comment (bassamsdata) [reply lua/codecompanion/config.lua]: Do you think it would be better to name the new tool `edit_file`, like other CLI agents do? Calling it `insert_edit_into_file` might mislead some LLMs, since the tool now does more than just insert edits into files.

File: tests/strategies/chat/tools/catalog/insert_edit_into_file/test_edit_file.lua (hunk 3 above)
Comment (olimorris) [reply tests/strategies/chat/tools/catalog/insert_edit_into_file/test_edit_file.lua:42]: Had a play at doing this. Turns out it's a real anti-pattern given we have to supply the edits to the tool as well. Feels like more work. I did refactor the tests to use `vim.fn.tempname()` though

**Comment by bassamsdata** [comment]
Next Steps

The core features of the tool are now complete 
Here’s what’s next:
	•	More testing, especially for edge and real-world cases (currently writing these).
	•	Refactor the file structure, it’s time for its own dedicated directory, I think?.
	•	Clean up the code a lot and document the tool at the top of the file for future enahncements.

Complementary Features
	•	Document formatting (default: false): this will be very useful once "agent" mode is live, helping detect syntax errors automatically.
	•	LSP diagnostics feedback (default: false): another step toward making the tool "agent"-ready in the future.

**Comment by olimorris** [comment]
Awesome work 👏🏼 .

Agree regarding its own folder. I think its worthwhile moving the helper files into that folder too unless you think they'll be used elsewhere in the plugin.

I'll take a thorough look in the coming days. One observation, we use `log:debug` a lot - Nearly 100 times in `edit_tool_exp_strategies`. Not that there's a magic number we should aim for but we'll be ones trawling through someones log file if they have an issue and that's a lot of information to process. Ideally `debug` will give us enough information to understand the process flow. Ideally `log:trace` would be used if we need really granular logging.

**Comment by olimorris** [comment]
On a slight aside, have you thought about writing a blog post on this work to share with the world? It's never been easier to send context to an LLM and receive a diff back...but applying it?! This is arguably one of the most challenging aspects of working directly with LLMs right now and it's been a huge source of frustration on my part.

**Comment by bassamsdata** [comment]
Absolutely, about the debug logs, I only added those for perosnal testing and debugging purposes. The code still needs a major cleanup, and I’ll probably start working on that in the next few days.

As for the blog post, I really like that idea, especially since I’ve been meaning to document how this tool works. I was originally planning to either write a detailed README in the tool’s directory or include a big block comment in the main Lua file, but turning it into a blog post sounds even better. 

When I first started, I couldn’t find any solid explanations about this topic (except docs from openai), most of what I came across were low-effort AI-generated _slop_ "articles". I had to learn by reading source code, exploring different projects written in various languages, and experimenting with real examples to see how each LLM would respond to different prompts or tool designs.

Once this gets merged, I’ll definitely document everything properly and share it here.

**Comment by olimorris** [comment]
It's really odd because no one seems to be willing to share their IP for how they accomplish it. There was a PR to CodeCompanion some months back where someone wanted to incorporate a third party service into the `insert_edit_into_file` tool. Would have improved it considerably but I didn't think it made sense to force users to have _another_ API key for a core tool, and send their data outside of the plugin and their LLM. But...it showed me that someone obviously thinks there's money to be made in that space.

**Comment by bassamsdata** [comment]
Hi @olimorris, hope you're doing well

Apologies for the delay in updating this PR. I’ve cleaned and organized the tool today, and it's nearly ready, I just need to do a bit more prompt testing for the new refactored prompt.
Before I mark it as "ready for review" I wanted to check in with you on the naming. Would you still prefer to stick with `insert_edit_into_file`, or would `edit_file` be a better fit? Personally, I lean toward `edit_file` since it more clearly reflects what the tool does.
I understand this might be a breaking change, so I’ll leave it up to you, I'm happy to update it either way, right away.

Thanks

**Comment by olimorris** [comment]
Please don't feel any pressure regarding any PR for CodeCompanion. Time is always going to be our enemy and we should prioritize life before this plugin 😄.

> Would you still prefer to stick with insert_edit_into_file

Yes. I want to continue to try and have parity between Copilot and CodeCompanion regarding tools, where possible. Also, it's one less thing to notify and potentially annoy some users with.

**Comment by olimorris** [comment]
Hey @bassamsdata, you happy for me to review this? I've blocked out some time tonight if you're happy with it.

**Comment by bassamsdata** [comment]
Hey @olimorris,

This is ready for review! Functionality-wise, I’m happy with how it’s working. There might be a few small prompt tweaks needed later on, but only if any issues come up.

There’s a minor issue with the winbar I just noticed, that I’ll try to sort out while you’re reviewing.

One quick note on the code, there are a few strategies we might eventually merge or even delete, but I think it’s best to keep them for now so we can gather some initial user feedback first.

Thanks

File: lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/strategies.lua [hunk 4]
```diff
@@ -0,0 +1,1062 @@
...
+  end
+
+  local matches = {}
+
+  if #search_lines == 0 then
+    return matches
+  end
+
+  -- Performance safeguards to prevent freezes
+  local max_content_lines = 5000
```
Comment (olimorris) [inline lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/strategies.lua]: As these are constants, propose putting them at the top level as:

```lua
local CONSTANTS = {
  MAX_CONTENT_LINES = 5000,
  MAX_SEARCH_LINES = 200,
  MAX_ITERATIONS = 10000,
}
```

To match the sort of pattern we have across other files

File: lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua [hunk 5]
```diff
@@ -0,0 +1,972 @@
...
+  -- Handle case where LLM sends edits as stringified JSON
+  if type(action.edits) == "string" then
+    local ok, parsed = pcall(vim.json.decode, action.edits)
+    if ok and type(parsed) == "table" then
+      action.edits = parsed
+    end
+  end
+
+  -- Early size validation to prevent freezes
+  if #current_content > 2000000 then -- 2MB limit
```
Comment (olimorris) [inline lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua]: Propose we put this as a constant at the top of the file like `local MAX_FILE_SIZE_BYTES = 2000000 -- 2MB`. At some point we may wish to make this a tool config option

File: lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/strategies.lua [hunk 6]
```diff
@@ -0,0 +1,1062 @@
+--[[
```
Comment (olimorris) [inline lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/strategies.lua:1]: There's quite a few luals warnings in this file:

- `normalize_whitespace` is defined twice
- `gsub` actually returns [two values](https://www.lua.org/pil/20.1.html) (the replaced string and the number of replacements) so we should do something like:

```lua
local replaced, _ = vim.trim(text):gsub("%s+", " ")
return replaced
```

File: lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua [hunk 7]
```diff
@@ -0,0 +1,972 @@
...
+      strict = true,
+    },
+  },
+  handlers = {
+    ---The handler to determine whether to prompt the user for approval
+    ---@param self CodeCompanion.Tool.EditFile
+    ---@param tools CodeCompanion.Tools
+    ---@param config table The tool configuration
+    ---@return boolean
+    prompt_condition = function(self, tools, config)
```
Comment (olimorris) [inline lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua]: We're overwriting the `config` at the top of the file. This was probably a legacy issue from my previous implementation

File: lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua [hunk 8]
```diff
@@ -0,0 +1,972 @@
...
+  local prompt_path = Path:new(dir, "prompt.md")
+  return prompt_path:read()
+end
+
+local PROMPT = load_prompt()
+
+-- Enhanced Python-like parser for handling LLM's mixed Python/JSON syntax
+---@param edits_string string The string containing edits in Python-like format
+---@return string The converted JSON-like string
+local function parse_python_like_edits(edits_string)
```
Comment (olimorris) [inline lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua]: `edits` is better than `edits_string` as we state the type is a string in the doc block above

File: lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/strategies.lua [hunk 9]
```diff
@@ -0,0 +1,1062 @@
...
+            total_attempts = vim.tbl_keys(all_strategies),
+            fallback_used = true,
+          }
+        else
+          goto continue
+        end
+      end
+
+      if selection_result.success then
+        log:debug("[Insert_edit_into_file Strategies] Strategy %s succeeded with confident match", strategy.name)
```
Comment (olimorris) [inline lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/strategies.lua]: I propose we think about a user who is having problems with the tool and they share their debug logs with us. Are we collecting too much data from them with this file and the init.lua file? Right now there's 32 log statements across the two files. Given these are disk I/O operations i'd advise trimming these down

File: lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/prompt.md [hunk 10]
```diff
@@ -0,0 +1,232 @@
+The insert_edit_into_file tool performs deterministic file edits using text-based matching. It supports exact, block, or substring replacements with optional context matching for safety.
```
Comment (olimorris) [inline lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/prompt.md:1]: Firstly, this is a wonderful document. Well thought out and I expect it's a core differentiator behind why this algorithm works so well.

File: lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/prompt.md [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+Use edit_tool_exp for safe, reliable file editing with advanced matching strategies.
+
+# Quick Reference
```
Comment (olimorris) [inline lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/prompt.md]: Propose we move these to H2 headers. We shouldn't have more than 1 H1 in a single file

File: lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua [hunk 12]
```diff
@@ -0,0 +1,972 @@
...
+  log:trace("[Insert_edit_into_file Tool] Trying enhanced Python-like parser")
+
+  -- Step 1: Fix boolean values first (before quote conversion)
```
Comment (olimorris) [inline lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua]: For some of these comments, I think it's worthwhile including the why. Your comprehensive testing has probably uncovered these so it's worthwhile documenting that. Something like:

```lua
-- Fix booleans before quote conversion to prevent 'True' → "True" string conversion
```

File: lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua
Comment (olimorris) [inline lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua:1]: The review point I missed...My preference is to have internal filepath parameters as `path` as it's just a more common Neovim/Vim term. However, externally facing (i.e. when talking to an LLM, keeping it as filepath. So things like `action.filepath` are totally fine but in `write_file_content` we should have that parameter as `path`.

I actually used the `insert_edit_into_file` tool to make those changes and it nailed it first time:

Image

**Comment by olimorris** [comment]
The other point I wanted to add was regarding the output messages in the chat buffer. With the last edit I made, I see:

```markdown
Applied 8 edit(s) to lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua
Rename all internal function parameters and variables from `filepath` to `path` while preserving `filepath` in the schema and user-facing error messages
```

For brevity, it would be cleaner if that first line read `Edited insert_edit_into_file/init.lua file` which, I recall is how Copilot Chat does it. As the user will likely have a diff to manually view the edits then the "applied" and "8 edits" is superfluous, saving us some space in the chat buffer. The reasoning message, however, is superb and it's great to see that.

**Comment by bassamsdata** [comment]
Sure, I’ll update that first line to say `Edited insert_edit_into_file/init.lua file`.

I’ll also tweak the `fmt()` output to clean up the English style (so we don’t get things like `1 edit(s)`). Adding this helper should do the trick in the dryRun ofcourse.

```lua
--- Return a word in singular or plural form based on count
function M.pluralize(count, word)
  return count == 1 and word or word .. "s"
end
```

**Comment by bassamsdata** [comment]
I’ve made several improvements, from message handling to debugging, to make the code more reliable, especially when working with smaller models. I spent the last couple of days testing with the 14B Qwen3 model and took the opportunity to refactor quite a bit. 

Things should now be easier to follow, maintain, and extend, with clearer separation of responsibilities.

**Review by olimorris — COMMENTED** [review]
This is amazing, @bassamsdata. Just super comprehensive, well thought out and I've not been able to break it 😆.

Comments below are super minor. I made them over the course of 12+ hours so they might not be very coherent so let me know if they don't make sense.

**Comment by olimorris** [comment]
Firstly, an apology. I didn't submit my preview on Sunday. So these comments have been sat here for a whole week as "pending". I'm an idiot.

My feedback is mostly minor, I did observe some LuaLs warnings but otherwise it's just nit picky stuff.

File: lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua (hunk 5 above)
Comment (bassamsdata) [reply lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua]: I’ve moved all constants into a separate file to make future updates and implementations easier and more organized.

Edit: added an option for size limit:
`file_size_limit_mb = 2, -- Maximum file size in MB`

**Comment by bassamsdata** [comment]
No worries at all! Most of the review comments were already addressed during my refactoring, and I’ve just resolved the remaining ones. I also added an option for the file size:

```lua
file_size_limit_mb = 2, -- Maximum file size in MB
```

Though worth mentioning, I also added two new files after your review, one for constants and another for `text_utils`, which I extracted from the strategies to make the code more maintainable.

Thanks again for the feedback!

**Comment by olimorris** [comment]
Awesome. Happy with those.

Shall we merge?

**Comment by olimorris** [comment]
One final thought. Is it worth including a sentence or two for a system prompt?

I added a memory tool last night. It's native to Anthropic so their models had no issue using it. But I had to prompt OpenAI quite heavily to use it. Only when I included a lite system prompt did they start using it like Anthropic.

**Comment by bassamsdata** [comment]
> One final thought. Is it worth including a sentence or two for a system prompt?

Done, I hope it’s not too much, but please let me know if it feels like too much.

> I added a memory tool last night.

Nice addition! When I rebased last night, I noticed it, though my late-night brain briefly thought it was the other kind of memory and wondered _'hmm, memory in the tools section!'_ 😅

I’m happy with the current state of the PR. If you’re good with it too, feel free to merge!

**Comment by olimorris** [comment]
Amazing. Thanks so much. I'll write the announcement now.

**Comment by graelo** [comment]
Thanks @bassamsdata that's brilliant work. Thanks @olimorris for everything. 🙇
